File: apps/shared/loopUnroller.ts
/**
 * 循环展开器 (Loop Unroller) - 共享模块
 * 
 * 将带有嵌套循环的工作流节点列表"展开"为实际执行步骤列表。
 * 
 * 用途：
 *   - 后端：按展开后的步骤顺序执行节点
 *   - 前端 ProgressBar：准确计算进度 (当前步骤 / 总步骤数)
 *   - 时间估算器：计算真实执行时间
 *   - 总结报告：统计实际执行次数
 *   - 断点续运行：定位到具体的迭代轮次
 * 
 * @module loopUnroller
 */

// =============================================================================
// 类型定义 (独立定义，不依赖具体项目的 Interfaces)
// =============================================================================

/**
 * 简化的节点类型（仅包含展开所需的字段）
 */
export interface SimpleNode {
    id: string;
    type: string;
    config?: Record<string, any>;
}

/**
 * 展开后的单个执行步骤
 */
export interface UnrolledStep {
    /** 原始节点 ID */
    nodeId: string;
    /** 节点类型 */
    nodeType: string;
    /** 在原始 nodes[] 数组中的索引位置 */
    originalIndex: number;
    /** 
     * 迭代路径 (数组形式，支持嵌套循环)
     * 
     * 示例：
     *   - [] 表示不在任何循环内
     *   - [2] 表示在第一层循环的第3轮 (0-based)
     *   - [1, 0] 表示在外层第2轮、内层第1轮
     */
    iterationPath: number[];
    /** 
     * 循环上下文栈 (记录所在的循环边界节点索引)
     * 
     * 示例：
     *   - [] 表示不在任何循环内
     *   - [0] 表示在 nodes[0] 开始的循环内
     *   - [0, 3] 表示在 nodes[0] 开始的循环内的 nodes[3] 开始的子循环内
     */
    loopContextStack: number[];
    /**
     * 循环深度 (0 = 不在循环内, 1 = 一层循环, ...)
     */
    loopDepth: number;
}

/**
 * 展开结果统计
 */
export interface UnrollSummary {
    /** 展开后的总步骤数 */
    totalSteps: number;
    /** 原始物理节点数 (不含 loop_start/loop_end) */
    physicalNodeCount: number;
    /** 最大嵌套深度 */
    maxLoopDepth: number;
    /** 循环信息列表 */
    loops: Array<{
        startIndex: number;
        endIndex: number;
        iterationCount: number;
        depth: number;
    }>;
}

/**
 * 完整的展开结果
 */
export interface UnrollResult {
    /** 展开后的步骤列表 */
    steps: UnrolledStep[];
    /** 统计摘要 */
    summary: UnrollSummary;
}

// =============================================================================
// 核心实现
// =============================================================================

/**
 * 查找匹配的 loop_end 节点索引 (支持嵌套循环)
 * 
 * @param nodes 节点列表
 * @param loopStartIdx loop_start 节点的索引
 * @returns loop_end 节点的索引，如果未找到返回 -1
 */
export function findMatchingLoopEnd(nodes: SimpleNode[], loopStartIdx: number): number {
    let depth = 0;
    for (let i = loopStartIdx; i < nodes.length; i++) {
        if (nodes[i].type === 'loop_start') depth++;
        if (nodes[i].type === 'loop_end') {
            depth--;
            if (depth === 0) return i;
        }
    }
    return -1;
}

/**
 * 递归展开循环
 * 
 * @param nodes 原始节点列表
 * @param startIdx 开始索引 (包含)
 * @param endIdx 结束索引 (不包含)
 * @param iterationPath 当前迭代路径
 * @param loopContextStack 当前循环上下文栈
 * @returns 展开后的步骤列表
 */
function unrollRecursive(
    nodes: SimpleNode[],
    startIdx: number,
    endIdx: number,
    iterationPath: number[],
    loopContextStack: number[]
): UnrolledStep[] {
    const steps: UnrolledStep[] = [];
    let i = startIdx;

    while (i < endIdx) {
        const node = nodes[i];

        // 情况 1：遇到 loop_start，递归展开循环体
        if (node.type === 'loop_start') {
            const loopCount = node.config?.loop_count ?? 1;
            const loopEndIdx = findMatchingLoopEnd(nodes, i);

            if (loopEndIdx === -1) {
                // 未找到匹配的 loop_end，静默跳过
                // 这在用户编辑工作流时是正常的（刚添加 loop_start，还没添加 loop_end）
                i++;
                continue;
            }

            // 对循环体执行 loopCount 次递归展开
            for (let iter = 0; iter < loopCount; iter++) {
                const childSteps = unrollRecursive(
                    nodes,
                    i + 1,            // 跳过 loop_start
                    loopEndIdx,       // 到 loop_end 之前
                    [...iterationPath, iter],
                    [...loopContextStack, i]
                );
                steps.push(...childSteps);
            }

            // 跳到 loop_end 之后
            i = loopEndIdx + 1;
        }
        // 情况 2：遇到 loop_end，直接跳过 (理论上不会走到这里)
        else if (node.type === 'loop_end') {
            i++;
        }
        // 情况 3：普通节点，添加到步骤列表
        else {
            steps.push({
                nodeId: node.id,
                nodeType: node.type,
                originalIndex: i,
                iterationPath: [...iterationPath],
                loopContextStack: [...loopContextStack],
                loopDepth: iterationPath.length
            });
            i++;
        }
    }

    return steps;
}

/**
 * 主函数：将工作流节点列表展开为执行步骤列表
 * 
 * @param nodes 原始工作流节点列表
 * @returns 展开结果 (包含步骤列表和统计摘要)
 * 
 * @example
 * ```typescript
 * const nodes = [
 *   { id: 'n1', type: 'loop_start', config: { loop_count: 3 } },
 *   { id: 'n2', type: 'eis_potentiostatic', config: {} },
 *   { id: 'n3', type: 'loop_end', config: {} }
 * ];
 * 
 * const result = unrollLoops(nodes);
 * console.log(result.summary.totalSteps); // 3
 * console.log(result.steps[0].iterationPath); // [0]
 * console.log(result.steps[1].iterationPath); // [1]
 * console.log(result.steps[2].iterationPath); // [2]
 * ```
 */
export function unrollLoops(nodes: SimpleNode[]): UnrollResult {
    // 执行递归展开
    const steps = unrollRecursive(nodes, 0, nodes.length, [], []);

    // 统计物理节点数 (排除 loop_start/loop_end)
    const physicalNodeCount = nodes.filter(
        n => n.type !== 'loop_start' && n.type !== 'loop_end'
    ).length;

    // 统计最大嵌套深度
    const maxLoopDepth = steps.reduce(
        (max, step) => Math.max(max, step.loopDepth),
        0
    );

    // 收集循环信息
    const loops: UnrollSummary['loops'] = [];
    for (let i = 0; i < nodes.length; i++) {
        if (nodes[i].type === 'loop_start') {
            const endIdx = findMatchingLoopEnd(nodes, i);
            if (endIdx !== -1) {
                // 计算当前循环的嵌套深度
                let depth = 0;
                for (let j = 0; j < i; j++) {
                    if (nodes[j].type === 'loop_start') depth++;
                    if (nodes[j].type === 'loop_end') depth--;
                }
                loops.push({
                    startIndex: i,
                    endIndex: endIdx,
                    iterationCount: nodes[i].config?.loop_count ?? 1,
                    depth
                });
            }
        }
    }

    return {
        steps,
        summary: {
            totalSteps: steps.length,
            physicalNodeCount,
            maxLoopDepth,
            loops
        }
    };
}

// =============================================================================
// 辅助工具函数
// =============================================================================

/**
 * 根据展开后的步骤索引，获取对应的原始节点信息
 * 
 * @param result 展开结果
 * @param stepIndex 展开后的步骤索引 (0-based)
 * @returns 步骤信息，如果索引越界返回 undefined
 */
export function getStepAt(result: UnrollResult, stepIndex: number): UnrolledStep | undefined {
    return result.steps[stepIndex];
}

/**
 * 查找原始节点在展开后出现的所有步骤索引
 * 
 * @param result 展开结果
 * @param originalIndex 原始节点索引
 * @returns 该节点在展开后对应的所有步骤索引
 */
export function findStepsByOriginalIndex(result: UnrollResult, originalIndex: number): number[] {
    return result.steps
        .map((step, idx) => step.originalIndex === originalIndex ? idx : -1)
        .filter(idx => idx !== -1);
}

/**
 * 格式化迭代路径为可读字符串
 * 
 * @param iterationPath 迭代路径数组
 * @returns 格式化字符串，如 "1-2-3" 表示第1轮→第2轮→第3轮 (1-based 显示)
 */
export function formatIterationPath(iterationPath: number[]): string {
    if (iterationPath.length === 0) return '-';
    return iterationPath.map(i => i + 1).join('-');
}

/**
 * 计算工作流的进度百分比
 * 
 * @param result 展开结果
 * @param currentStepIndex 当前执行到的步骤索引 (展开后的索引)
 * @returns 进度百分比 (0-100)
 */
export function calculateProgress(result: UnrollResult, currentStepIndex: number): number {
    if (result.summary.totalSteps === 0) return 0;
    return Math.min(100, (currentStepIndex / result.summary.totalSteps) * 100);
}

File: apps/python_backend/loop_unroller.py
"""Loop unroller for the unified Python backend."""

from __future__ import annotations

from runtime.execution_semantics import ADVANCED_MEASUREMENT_TYPES, MEASUREMENT_NODE_TYPES

LOOP_START = "loop_start"
LOOP_END = "loop_end"
WORKFLOW_BLOCK = "workflow_block"
IGNORED_BLOCK_NODE_TYPES = {"startup", "shutdown"}
BOUNDARY_NODE_TYPES = {"startup", "shutdown"}


class WorkflowBlockError(ValueError):
    """Raised when a workflow block cannot be expanded for v1 execution."""


def find_matching_loop_end(nodes: list[dict], loop_start_idx: int, end_idx: int | None = None) -> int:
    depth = 0
    limit = len(nodes) if end_idx is None else min(end_idx, len(nodes))
    for index in range(loop_start_idx, limit):
        node_type = nodes[index].get("type")
        if node_type == LOOP_START:
            depth += 1
        elif node_type == LOOP_END:
            depth -= 1
            if depth == 0:
                return index
    return -1


def unroll_loops(nodes: list[dict], workflow_loader=None, auto_startup_config: dict | None = None) -> dict:
    steps = _unroll_recursive(nodes, 0, len(nodes), [], [], [], workflow_loader or _load_workflow_nodes)
    steps = _with_measurement_boundaries(steps, auto_startup_config or {})
    for unrolled_index, step in enumerate(steps):
        step["unrolledIndex"] = unrolled_index
        step["unrolledTotal"] = len(steps)
    return {"steps": steps, "summary": _build_summary(nodes, steps)}


def _unroll_recursive(
    nodes: list[dict],
    start_idx: int,
    end_idx: int,
    iteration_path: list[dict],
    loop_context_stack: list[int],
    block_path: list[dict],
    workflow_loader,
    parent_original_index: int | None = None,
) -> list[dict]:
    steps: list[dict] = []
    index = start_idx

    while index < end_idx:
        node = nodes[index]
        node_type = node.get("type")

        if node_type == LOOP_START:
            loop_end_idx = find_matching_loop_end(nodes, index, end_idx)
            if loop_end_idx == -1:
                index += 1
                continue

            total_iterations = _loop_count(node)
            body_node_indices = _body_node_indices(nodes, index + 1, loop_end_idx)
            for iteration_idx in range(total_iterations):
                loop_event = {
                    "loopStartIndex": index,
                    "iteration": iteration_idx + 1,
                    "totalIterations": total_iterations,
                    "nodeIndices": body_node_indices,
                }
                child_path = [
                    *iteration_path,
                    {
                        "loopNodeId": node.get("id", ""),
                        "loopStartIndex": index,
                        "iteration": iteration_idx + 1,
                        "totalIterations": total_iterations,
                    },
                ]
                child_steps = _unroll_recursive(
                    nodes,
                    index + 1,
                    loop_end_idx,
                    child_path,
                    [*loop_context_stack, index],
                    block_path,
                    workflow_loader,
                    parent_original_index,
                )
                if child_steps:
                    child_steps[0].setdefault("loopEvents", []).insert(0, loop_event)
                steps.extend(child_steps)

            index = loop_end_idx + 1
            continue

        if node_type == WORKFLOW_BLOCK:
            child_steps = _expand_workflow_block(
                node,
                index,
                iteration_path,
                loop_context_stack,
                block_path,
                workflow_loader,
                parent_original_index,
            )
            steps.extend(child_steps)
            index += 1
            continue

        if node_type in ADVANCED_MEASUREMENT_TYPES:
            steps.extend(
                _expand_advanced_node(
                    node,
                    index,
                    iteration_path,
                    loop_context_stack,
                    block_path,
                    parent_original_index,
                )
            )
            index += 1
            continue

        if node_type == LOOP_END:
            index += 1
            continue

        steps.append(
            {
                "nodeId": node.get("id"),
                "nodeType": node_type,
                "originalIndex": parent_original_index if parent_original_index is not None else index,
                "sourceIndex": index,
                "node": node,
                "iterationPath": [dict(item) for item in iteration_path],
                "loopContextStack": list(loop_context_stack),
                "loopDepth": len(loop_context_stack),
                "blockPath": [dict(item) for item in block_path],
            }
        )
        index += 1

    return steps


def _expand_advanced_node(
    node: dict,
    index: int,
    iteration_path: list[dict],
    loop_context_stack: list[int],
    block_path: list[dict],
    parent_original_index: int | None,
) -> list[dict]:
    config = dict(node.get("config") or {})
    node_type = node.get("type")
    original_index = parent_original_index if parent_original_index is not None else index
    common_meta = {
        "parentNodeId": node.get("id"),
        "parentNodeType": node_type,
        "nodeConfig": config,
    }

    if node_type in ("galvanostatic_step_ramp", "potentiostatic_step_ramp"):
        is_galvanostatic = node_type == "galvanostatic_step_ramp"
        start = _float_config(
            config,
            ["startCurrent", "start_current", "start_potential", "startPotential"],
            0.1 if is_galvanostatic else 0.0,
        )
        end = _float_config(config, ["endCurrent", "end_current", "end_potential", "endPotential"], 1.0)
        step_value = _float_config(config, ["stepCurrent", "step_current", "step_potential", "stepPotential"], 0.1)
        hold_time = _float_config(config, ["holdTime", "hold_time", "measurementDuration"], 30.0)
        sampling_interval = _float_config(config, ["samplingInterval", "sampling_interval"], 0.5)
        if not step_value:
            step_value = 0.1
        raw_step_count = int(abs(end - start) // abs(step_value)) + 1
        step_count = min(max(1, raw_step_count), 1000)
        direction = 1 if end >= start else -1
        actual_step = abs(step_value) * direction
        child_type = "chronopotentiometry" if is_galvanostatic else "chronoamperometry"
        child_steps = []

        for step_index in range(step_count):
            value = start + step_index * actual_step
            child_config = {
                **config,
                **common_meta,
                "measurementDuration": hold_time,
                "samplingInterval": sampling_interval,
                "stepIndex": step_index,
                "totalSteps": step_count,
                "stepValue": value,
            }
            if is_galvanostatic:
                child_config["polarizationCurrent"] = value
                child_config["polarization_current"] = value
            else:
                child_config["polarizationVoltage"] = value
                child_config["polarization_voltage"] = value
            child_steps.append(
                _advanced_child_step(
                    node,
                    child_type,
                    child_config,
                    original_index,
                    step_index,
                    iteration_path,
                    loop_context_stack,
                    block_path,
                )
            )
        return child_steps

    if node_type in ("galvanostatic_switching", "potentiostatic_switching"):
        is_galvanostatic = node_type == "galvanostatic_switching"
        cycles = max(0, int(_float_config(config, ["cycles"], 5)))
        sampling_interval = _float_config(config, ["samplingInterval", "sampling_interval"], 0.5)
        first_value = _float_config(
            config,
            ["current1", "current_1", "potential1", "potential_1"],
            0.0,
        )
        second_value = _float_config(
            config,
            ["current2", "current_2", "potential2", "potential_2"],
            0.01 if is_galvanostatic else 0.5,
        )
        hold_times = [
            _float_config(config, ["holdTime1", "hold_time_1"], 30.0),
            _float_config(config, ["holdTime2", "hold_time_2"], 30.0),
        ]
        values = [first_value, second_value]
        child_type = "chronopotentiometry" if is_galvanostatic else "chronoamperometry"
        total_steps = cycles * 2
        child_steps = []

        for cycle_index in range(cycles):
            for phase_index, value in enumerate(values):
                step_index = cycle_index * 2 + phase_index
                child_config = {
                    **config,
                    **common_meta,
                    "measurementDuration": hold_times[phase_index],
                    "samplingInterval": sampling_interval,
                    "stepIndex": step_index,
                    "totalSteps": total_steps,
                    "cycleIndex": cycle_index,
                    "isFirstOfCycle": phase_index == 0,
                    "stepValue": value,
                }
                if is_galvanostatic:
                    child_config["polarizationCurrent"] = value
                    child_config["polarization_current"] = value
                else:
                    child_config["polarizationVoltage"] = value
                    child_config["polarization_voltage"] = value
                child_steps.append(
                    _advanced_child_step(
                        node,
                        child_type,
                        child_config,
                        original_index,
                        step_index,
                        iteration_path,
                        loop_context_stack,
                        block_path,
                    )
                )
        return child_steps

    return []


def _advanced_child_step(
    parent_node: dict,
    child_type: str,
    child_config: dict,
    original_index: int,
    step_index: int,
    iteration_path: list[dict],
    loop_context_stack: list[int],
    block_path: list[dict],
) -> dict:
    parent_id = parent_node.get("id", "advanced")
    child_node = {
        "id": f"{parent_id}__expanded_{step_index + 1}",
        "type": child_type,
        "config": child_config,
    }
    return {
        "nodeId": child_node["id"],
        "nodeType": child_type,
        "originalIndex": original_index,
        "sourceIndex": original_index,
        "node": child_node,
        "iterationPath": [dict(item) for item in iteration_path],
        "loopContextStack": list(loop_context_stack),
        "loopDepth": len(loop_context_stack),
        "blockPath": [dict(item) for item in block_path],
        "parentNodeId": parent_id,
        "parentNodeType": parent_node.get("type"),
        "stepIndex": step_index,
        "totalSteps": child_config.get("totalSteps"),
        "stepValue": child_config.get("stepValue"),
        "cycleIndex": child_config.get("cycleIndex"),
    }


def _expand_workflow_block(
    node: dict,
    index: int,
    iteration_path: list[dict],
    loop_context_stack: list[int],
    block_path: list[dict],
    workflow_loader,
    parent_original_index: int | None,
) -> list[dict]:
    config = node.get("config") or {}
    workflow_id = str(config.get("workflowId") or "").strip()
    if not workflow_id:
        raise WorkflowBlockError("Workflow block is missing workflowId")

    workflow = workflow_loader(workflow_id)
    child_nodes = list((workflow or {}).get("nodes") or [])
    nested_block = next((child for child in child_nodes if child.get("type") == WORKFLOW_BLOCK), None)
    if nested_block:
        raise WorkflowBlockError("Nested workflow blocks are not supported in v1")

    executable_nodes = [
        child
        for child in child_nodes
        if child.get("type") not in IGNORED_BLOCK_NODE_TYPES
    ]
    child_block_path = [
        *block_path,
        {
            "blockNodeId": node.get("id", ""),
            "blockWorkflowId": workflow_id,
            "blockWorkflowName": config.get("workflowName") or (workflow or {}).get("name") or workflow_id,
            "blockOriginalIndex": parent_original_index if parent_original_index is not None else index,
        },
    ]
    return _unroll_recursive(
        executable_nodes,
        0,
        len(executable_nodes),
        iteration_path,
        loop_context_stack,
        child_block_path,
        workflow_loader,
        parent_original_index if parent_original_index is not None else index,
    )


def _load_workflow_nodes(workflow_id: str) -> dict:
    from database import db
    import json

    row = db.conn.execute("SELECT json_data FROM workflows WHERE id = ?", (workflow_id,)).fetchone()
    if not row:
        raise WorkflowBlockError(f"Workflow block references missing workflow: {workflow_id}")
    return json.loads(row["json_data"])


def _with_measurement_boundaries(steps: list[dict], startup_config: dict) -> list[dict]:
    non_boundary_steps = [step for step in steps if step.get("nodeType") not in BOUNDARY_NODE_TYPES]
    measurement_positions = [
        index
        for index, step in enumerate(non_boundary_steps)
        if step.get("nodeType") in MEASUREMENT_NODE_TYPES
    ]
    if not measurement_positions:
        return steps

    first_measurement = non_boundary_steps[measurement_positions[0]]
    last_measurement_position = measurement_positions[-1]
    last_measurement = non_boundary_steps[last_measurement_position]
    with_boundaries = list(non_boundary_steps)
    with_boundaries.insert(
        measurement_positions[0],
        _auto_boundary_step(
            "startup",
            "__auto_startup_before_measurement",
            first_measurement,
            startup_config,
        ),
    )
    with_boundaries.insert(
        last_measurement_position + 2,
        _auto_boundary_step(
            "shutdown",
            "__auto_shutdown_after_measurement",
            last_measurement,
            {},
        ),
    )
    return with_boundaries


def _auto_boundary_step(node_type: str, node_id: str, anchor_step: dict, config: dict) -> dict:
    original_index = int(anchor_step.get("originalIndex") or 0)
    return {
        "nodeId": node_id,
        "nodeType": node_type,
        "originalIndex": original_index,
        "sourceIndex": anchor_step.get("sourceIndex", original_index),
        "node": {
            "id": node_id,
            "type": node_type,
            "config": dict(config),
            "auto": True,
        },
        "iterationPath": [],
        "loopContextStack": [],
        "loopDepth": 0,
        "blockPath": [],
        "autoBoundary": True,
    }


def _loop_count(node: dict) -> int:
    config = node.get("config") or {}
    raw_count = config.get("loopCount", 1)
    try:
        return max(0, int(raw_count))
    except (TypeError, ValueError):
        return 1


def _float_config(config: dict, keys: list[str], default: float) -> float:
    for key in keys:
        if key not in config:
            continue
        try:
            return float(config[key])
        except (TypeError, ValueError):
            continue
    return float(default)


def _body_node_indices(nodes: list[dict], start_idx: int, end_idx: int) -> list[int]:
    return [
        index
        for index in range(start_idx, end_idx)
        if nodes[index].get("type") not in (LOOP_START, LOOP_END)
    ]


def _build_summary(nodes: list[dict], steps: list[dict]) -> dict:
    loops = []
    for index, node in enumerate(nodes):
        if node.get("type") != LOOP_START:
            continue
        loop_end_idx = find_matching_loop_end(nodes, index)
        if loop_end_idx == -1:
            continue
        loops.append(
            {
                "startIndex": index,
                "endIndex": loop_end_idx,
                "iterationCount": _loop_count(node),
                "depth": _loop_depth_at(nodes, index),
            }
        )

    return {
        "totalSteps": len(steps),
        "physicalNodeCount": len([node for node in nodes if node.get("type") not in (LOOP_START, LOOP_END)]),
        "maxLoopDepth": max((int(step.get("loopDepth") or 0) for step in steps), default=0),
        "loops": loops,
    }


def _loop_depth_at(nodes: list[dict], target_idx: int) -> int:
    depth = 0
    for index in range(target_idx):
        node_type = nodes[index].get("type")
        if node_type == LOOP_START and find_matching_loop_end(nodes, index) > target_idx:
            depth += 1
        elif node_type == LOOP_END:
            depth = max(0, depth - 1)
    return depth
